crates/common/src/raindex_client/local_db/fetch.rs

```diff
@@ -119,14 +119,108 @@ impl LocalDb {
         all_events.sort_by(|a, b| {
             let block_a = extract_block_number_from_entry(a).unwrap_or(0);
             let block_b = extract_block_number_from_entry(b).unwrap_or(0);
-            block_a.cmp(&block_b)
+            block_a
+                .cmp(&block_b)
+                .then_with(|| a.log_index.cmp(&b.log_index))
         });
 
         self.backfill_missing_timestamps(&mut all_events, config)
             .await?;
         Ok(all_events)
     }
 
+    pub async fn fetch_store_set_events(
+        &self,
+        store_addresses: &[String],
+        start_block: u64,
+        end_block: u64,
+        config: &FetchConfig,
+    ) -> Result<Vec<LogEntryResponse>, LocalDbError> {
+        if store_addresses.is_empty() {
+            return Ok(Vec::new());
+        }
+
+        let unique_addresses: Vec<String> = {
+            let mut dedup = HashSet::new();
+            store_addresses
+                .iter()
+                .filter_map(|addr| {
+                    let lower = addr.to_ascii_lowercase();
+                    if dedup.insert(lower.clone()) {
+                        Some(lower)
+                    } else {
+                        None
+                    }
+                })
+                .collect()
+        };
+
+        if unique_addresses.is_empty() {
+            return Ok(Vec::new());
+        }
+
+        let mut jobs = Vec::new();
+        let chunk_size = config.chunk_size.max(1);
+        let chunk_span = chunk_size.saturating_sub(1);
+        for address in unique_addresses.into_iter() {
+            let mut current_block = start_block;
+            while current_block <= end_block {
+                let to_block = current_block.saturating_add(chunk_span).min(end_block);
+                jobs.push((address.clone(), current_block, to_block));
+                current_block = to_block.saturating_add(1);
+                if to_block == u64::MAX {
+                    break;
+                }
+            }
+        }
```

Comment:
this doesn't really make it more concise but also an option if you like it

```suggestion
        let chunk_size = config.chunk_size.max(1);
        let chunk_span = chunk_size.saturating_sub(1);
        let jobs: Vec<_> = unique_addresses
            .into_iter()
            .flat_map(|address| {
                std::iter::successors(Some(start_block), |&current| {
                    if current >= end_block || current == u64::MAX {
                        None
                    } else {
                        Some(current.saturating_add(chunk_span).min(end_block).saturating_add(1))
                    }
                })
                .map(move |current_block| {
                    let to_block = current_block.saturating_add(chunk_span).min(end_block);
                    (address.clone(), current_block, to_block)
                })
            })
            .collect();
```